### PR TITLE
Separate guest/host tracking + unaligned protection

### PR DESCRIPTION
**Pull request by riperiperi**
This PR allows guest memory access and managed memory access (typically from the GPU processing thread) to have different memory tracking handles and protections. 

- Protection regions can be different for guest access. This is useful for the new host tracked memory manager in 16k page mode, as memory protection there only counts the first byte of the access. This causes issues when accesses are unaligned and cross over a page boundary, as it's possible to write data from the previous page without triggering pending actions or signalling the write flag. The solution here is to extend page protections to the previous page, but only for resources that require it, and only for guest accesses.
  - This PR also adds the code that enables this; you can specify that a memory has "single byte guest tracking", and you can specify that specific resources can be written with unaligned access.
  - Using this mode requires that the memory manager gracefully handle protecting (partially) unmapped pages, without dropping protection on the overlapped pages that are mapped. The host tracked memory manager might need to change to support this.
- This could be used to register actions that only happen on the guest, or only happen from managed threads. For example, flush actions for buffers need to flush all pending data from the GPU thread, but when triggered from the guest it only needs to flush data that the guest should be able to see at the time.
  - This isn't done right now, but will be useful in the future.
- Certain resources could entirely skip guest memory tracking, and only listen for internal access. 
  - This would be useful for host imported buffers in the future. Reads from the GPU thread would still need to wait for sync to ensure data is available, and writes would indicate that the buffer needs to have a shadow copy.

This PR shouldn't change existing functionality or performance, it's just setting up for things that will be done in the future. It needs sanity testing to make sure complex games still function, as it does change a core part of the emulator.

Software, host mapped and hypervisor memory modes should be tested to ensure they still work properly.

### CODE REVIEW

File: src/Ryujinx.Cpu/Jit/MemoryManager.cs [hunk 1]
```diff
@@ -573,7 +616,7 @@ public SmartMultiRegionHandle BeginSmartGranularTracking(ulong address, ulong si
         }
 
         /// <inheritdoc/>
```
Comment (gdkchan) [inline src/Ryujinx.Cpu/Jit/MemoryManager.cs:618]: I guess the inheritdoc is broken now that the method name changed?

File: src/Ryujinx.Graphics.Gpu/Memory/PhysicalMemory.cs [hunk 2]
```diff
@@ -408,12 +409,13 @@ public GpuRegionHandle BeginTracking(MultiRange range, ResourceKind kind)
         /// <param name="address">CPU virtual address of the region</param>
         /// <param name="size">Size of the region</param>
         /// <param name="kind">Kind of the resource being tracked</param>
+        /// <param name="flags">Region flags</param>
         /// <param name="handles">Handles to inherit state from or reuse</param>
         /// <param name="granularity">Desired granularity of write tracking</param>
         /// <returns>The memory tracking handle</returns>
-        public MultiRegionHandle BeginGranularTracking(ulong address, ulong size, ResourceKind kind, IEnumerable<IRegionHandle> handles = null, ulong granularity = 4096)
+        public MultiRegionHandle BeginGranularTracking(ulong address, ulong size, ResourceKind kind, RegionFlags flags = RegionFlags.None, IEnumerable<IRegionHandle> handles = null, ulong granularity = 4096)
```
Comment (gdkchan) [inline src/Ryujinx.Graphics.Gpu/Memory/PhysicalMemory.cs:416]: Can you place each parameter on a separate line, now that this line is very long?

File: src/Ryujinx.Memory/Tracking/MemoryTracking.cs [hunk 3]
```diff
@@ -27,16 +32,27 @@ public class MemoryTracking
         /// Create a new tracking structure for the given "physical" memory block,
         /// with a given "virtual" memory manager that will provide mappings and virtual memory protection.
         /// </summary>
+        /// <remarks>
+        /// If singleByteGuestTracking is true, the memory manager must also support protection on partially
```
Comment (gdkchan) [inline src/Ryujinx.Memory/Tracking/MemoryTracking.cs:36]: Nit: could use `paramref` for `singleByteGuestTracking` reference.

**Comment by MutantAura** [comment]
Should this be tested alongside https://github.com/Ryujinx/Ryujinx/pull/6356 or is it considered standalone.

**Comment by LukeWarnut** [comment]
Tested Splatoon 3, Splatoon 2, DOOM 2016, and Pokemon Violet with host unchecked. No regressions found.

Tested Mario vs. Donkey Kong Demo and Jurassic Park Classic Games Collection with all 3 modes. No regressions found.

**Comment by riperiperi** [comment]
#6356 will need modification after this is merged to support the separate tracking. I have a commit here that shows the basics:
https://github.com/riperiperi/Ryujinx/tree/feature/separate-tracking-ht

File: src/Ryujinx.Cpu/Jit/MemoryManager.cs (hunk 1 above)
Comment (riperiperi) [reply src/Ryujinx.Cpu/Jit/MemoryManager.cs:618]: Yeah. `Impl`s don't have docs in these files so I'll just remove it.

**Review by gdkchan — APPROVED** [review]
lgtm, thanks for making a path forward for the host tracked memory manager!

**Pull request merged**